docs: add documentation around importing CRD as GVK (#321)
defenseunicorns/pepr: docs/customresources.md

# Importing Custom Resources

> The [Kubernetes Fluent Client](https://github.com/defenseunicorns/kubernetes-fluent-client), empowers us to create TypeScript typings directly from Kubernetes Custom Resource Definitions (CRDs), resulting in file outputs. These generated files can be seamlessly incorporated into our Pepr capabilities, providing us with the ability to work with types derived from the CRDs. For instance, we can import Istio CRDs and manipulate them as though they were intrinsic Kubernetes resources.

Reviewer: ```suggestion
The [Kubernetes Fluent Client](https://github.com/defenseunicorns/kubernetes-fluent-client) supports the creation of TypeScript typings directly from Kubernetes Custom Resource Definitions (CRDs).  The files it generates can be directly incorporated into Pepr capabilities and provide a way to work with strongly-typed CRDs.

For example (below), Istio CRDs can be imported and used as though they were intrinsic Kubernetes resources.
```
Rework to remove "us" / "our" language to better match the rest of the write-up.

## Generating TypeScript Types from CRDs

> To produce a new type, employ the kubernetes-fluent-client CLI.

Reviewer: ```suggestion
Using the kubernetes-fluent-client to produce a new type looks like this:
```
Made it read like a lead-in for the example box below since, otherwise, it just restates what we said in the first paragraph _and_ the section header.

```bash
npx kubernetes-fluent-client crd [source] [directory]
```

> The `crd` commands expects a `[source]`, which can be a URL or local file containing the `CustomResourceDefinition(s)`, and a `[directory]` where the generated code will live.

Reviewer: ```suggestion
The `crd` command expects a `[source]`, which can be a URL or local file containing the `CustomResourceDefinition(s)`, and a `[directory]` where the generated code will live.
```

The following example creates types for the Istio CRDs:

```bash
┌─[cmwylie19@Cases-MacBook-Pro] - [/istio-integration] - [2023-10-16 05:46:24]

> └─[0] <> npx kubernetes-fluent-client crd https://raw.githubusercontent.com/istio/istio/master/manifests/charts/base/crds/crd-all.gen.yaml crds

Reviewer: ```suggestion
user@workstation$  npx kubernetes-fluent-client crd https://raw.githubusercontent.com/istio/istio/master/manifests/charts/base/crds/crd-all.gen.yaml crds
```

Not a huge deal, but your terminal prompt is loooooonnngggg and includes your username / host.  I'd probably shorten & anonymize if it was me, but really up to you.

- Generating networking.istio.io/v1beta1 types for ProxyConfig
- Generating networking.istio.io/v1alpha3 types for ServiceEntry
- Generating networking.istio.io/v1beta1 types for ServiceEntry
- Generating networking.istio.io/v1alpha3 types for Sidecar
- Generating networking.istio.io/v1beta1 types for Sidecar
- Generating networking.istio.io/v1alpha3 types for VirtualService
- Generating networking.istio.io/v1beta1 types for VirtualService
- Generating networking.istio.io/v1alpha3 types for WorkloadEntry
- Generating networking.istio.io/v1beta1 types for WorkloadEntry
- Generating networking.istio.io/v1alpha3 types for WorkloadGroup
- Generating networking.istio.io/v1beta1 types for WorkloadGroup
- Generating security.istio.io/v1 types for AuthorizationPolicy
- Generating security.istio.io/v1beta1 types for AuthorizationPolicy
- Generating security.istio.io/v1beta1 types for PeerAuthentication
- Generating security.istio.io/v1 types for RequestAuthentication
- Generating security.istio.io/v1beta1 types for RequestAuthentication
- Generating telemetry.istio.io/v1alpha1 types for Telemetry

✅ Generated 23 files in the istio directory
```

> We observe that the `kubernetes-fluent-client` has produced the TypeScript types within the `crds` directory. These types can now be utilized in the Pepr module.

Reviewer: ```suggestion
Observe that the `kubernetes-fluent-client` has produced the TypeScript types within the `crds` directory. These types can now be utilized in the Pepr module.
```

```typescript
┌─[cmwylie19@Cases-MacBook-Pro] - [/istio-integration] - [2023-10-16 05:46:38]

> └─[0] <> cat crds/proxyconfig-v1beta1.ts

Reviewer: ```suggestion
user@workstation$  cat crds/proxyconfig-v1beta1.ts
```
Or whatever you prefer.

     * https://istio.io/docs/reference/config/networking/proxy-config.html
     */
    spec?:   Spec;
    status?: { [key: string]: any };
}

/**
 * Provides configuration for individual workloads. See more details at:
 * https://istio.io/docs/reference/config/networking/proxy-config.html
 */
export interface Spec {
    /**
     * The number of worker threads to run.
     */
    concurrency?: number;
    /**
     * Additional environment variables for the proxy.
     */
    environmentVariables?: { [key: string]: string };
    /**
     * Specifies the details of the proxy image.
     */
    image?: Image;
    /**
     * Optional.
     */
    selector?: Selector;
}

/**
 * Specifies the details of the proxy image.
 */
export interface Image {
    /**
     * The image type of the image.
     */
    imageType?: string;
}

/**
 * Optional.
 */
export interface Selector {
    /**
     * One or more labels that indicate a specific set of pods/VMs on which a policy should be
     * applied.
     */
    matchLabels?: { [key: string]: string };
}

RegisterKind(ProxyConfig, {
  group: "networking.istio.io",
  version: "v1beta1",
  kind: "ProxyConfig",
});
```

## Using new types

> The generated types can be imported into Pepr, _there is no additional logic added for this to work_. 

Reviewer: ```suggestion
The generated types can be imported into Pepr directly, _there is no additional logic needed to make them to work_. 
```
Reads more like rest of doc.
